feat: oci logging analytics output plugin implementation

adiforluls opened the pull request:
<!-- Provide summary of changes -->
A new output plugin for OCI Logging Analytics. 
https://docs.oracle.com/en-us/iaas/logging-analytics/index.html
REST API reference of the output destination: https://docs.oracle.com/en-us/iaas/api/#/en/logan-api-spec/20200601/Upload/UploadLogEventsFile

<!-- Issue number, if available. E.g. "Fixes #31", "Addresses #42, #77" -->

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [x] Example configuration file for the change
- [x] Debug log output from testing the change
<!--  
Please refer to the Developer Guide for instructions on building Fluent Bit with Valgrind support: 
https://github.com/fluent/fluent-bit/blob/master/DEVELOPER_GUIDE.md#valgrind
Invoke Fluent Bit and Valgrind as: $ valgrind --leak-check=full ./bin/fluent-bit <args>
-->
- [ ] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

If this is a change to packaging of containers or native binaries then please confirm it works for all targets.
- [ ] Run [local packaging test](./packaging/local-build-all.sh) showing all targets (including any new ones) build.
- [ ] Set `ok-package-test` label to test for all targets (requires maintainer to do).

**Documentation**
<!-- Docs can be edited at https://github.com/fluent/fluent-bit-docs -->
- [x] Documentation required for this feature

<!--  Doc PR (not required but highly recommended) -->
https://github.com/fluent/fluent-bit-docs/pull/1186

**Backporting**
<!--
PRs targeting the default master branch will go into the next major release usually.
If this PR should be backported to the current or earlier releases then please submit a PR for that particular branch.
-->
- [ ] Backport to latest stable release.

<!--  Other release PR (not required but highly recommended for quick turnaround) -->
----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

adiforluls commented:
```
[Output]
    Name oci_logan
    Match *
    Namespace <REDACTED>
    config_file_location <absolute-path-to-oci-config-file>
    profile_name ADMIN
    oci_la_log_group_id ocid1.loganalyticsloggroup.oc1<REDACTED>tkovmrkq
    oci_la_log_source_name Kubernetes Container Generic Logs
    tls On
    tls.verify Off
```
```
[2023/08/15 18:36:22] [ info] [output:oci_logan:oci_logan.0] payload={"logEvents":[{"logSourceName":"Kubernetes Container Generic Logs","logRecords":["I0801 00:19:20.282516      12 store.go:588] \"Secret was added and it is used in ingress annotations. Parsing\" secret=\"cattle-system/tls-rancher-ingress\"","I0801 00:19:20.283998      12 backend_ssl.go:67] \"Adding secret to local store\" name=\"cattle-system/tls-rancher-ingress\"","W0801 00:19:21.667885      12 controller.go:1152] Service \"cattle-system/rancher\" does not have any active Endpoint.","I0801 00:19:21.668014      12 controller.go:190] \"Configuration changes detected, backend reload required\"","I0801 00:19:21.718198      12 controller.go:207] \"Backend successfully reloaded\""]}]}
[2023/08/15 18:36:22] [ info] [output:oci_logan:oci_logan.0] lg_id=ocid1.loganalyticsloggroup.oc1.iad.amaaaaaalknnwbqaldbpnvd5mjs6tjyhprh7s43ftwk2efoo47ggtkovmrkq
[2023/08/15 18:36:22] [ info] [output:oci_logan:oci_logan.0] full_uri=/20200601/namespaces/odsbuilddev/actions/uploadLogEventsFile?logGroupId=ocid1.loganalyticsloggroup.oc1.iad.amaaaaaalknnwbqaldbpnvd5mjs6tjyhprh7s43ftwk2efoo47ggtkovmrkq&payloadType=JSON
[2023/08/15 18:36:23] [ info] [output:oci_logan:oci_logan.0] built client
[2023/08/15 18:36:23] [ info] [output:oci_logan:oci_logan.0] built request
[2023/08/15 18:36:23] [ info] [output:oci_logan:oci_logan.0] request header POST /20200601/namespaces/odsbuilddev/actions/uploadLogEventsFile?logGroupId=ocid1.loganalyticsloggroup.oc1.iad.amaaaaaalknnwbqaldbpnvd5mjs6tjyhprh7s43ftwk2efoo47ggtkovmrkq&payloadType=JSON HTTP/1.1

[2023/08/15 18:36:24] [ info] [output:oci_logan:oci_logan.0] placed request
[2023/08/15 18:36:24] [ info] [output:oci_logan:oci_logan.0] success

```

adiforluls commented:
To code-owners: PR description includes the link of REST API that is used to flush logs. The authorization is done using a configuration file, following is the documentation for that https://docs.oracle.com/en-us/iaas/Content/API/Concepts/sdkconfig.htm

adiforluls commented:
I'll fix the merge conflict and include appropriate copyrights tomorrow.

adiforluls commented:
cc: @edsiper @leonardo-albertovich

adiforluls commented:
Sure @patrick-stephens, I've been able to build it on macOS (the log output is from a binary built and run on macOS).

Also I don't see that tests ran for this PR? All I see are a bunch of tests skipped

patrick-stephens commented:
Yeah Github status is not reporting any problems but it has not triggered the CI for package builds for some reason or any of the unit tests.

patrick-stephens commented:
Ah @adiforluls the reason is you have conflicts that need to be resolved - nothing can run until that is sorted.

patrick-stephens commented:
Yeah, you can see there are a few coding issues, e.g. on older targets this is not valid syntax:
```
#11 99.52 /src/plugins/out_oci_logan/oci_logan.c:1115:9: error: 'for' loop initial declarations are only allowed in C99 mode
#11 99.52          for(int i = 0; i < map_size; i++) {
```
Check the coding style but I think Fluent Bit requires everything pre-declared.

adiforluls commented:
Fixed coding style and included appropriate copyrights. I see some macOS tests failing, I'll take a look if there's a problem tomorrow.

adiforluls commented:
The two macOS tests that failed, seeing that they have failed in other pull requests + commits that went into master, so assuming it has nothing to do with the changes here. If there's something that might be related let me know.

@patrick-stephens @edsiper @leonardo-albertovich et al, requesting a code review if there are no problems.

edsiper commented on plugins/out_oci_logan/oci_logan.c:
```diff
@@ -0,0 +1,1327 @@
+{
+    flb_sds_t content;
+
+    content = flb_sds_create_size(512);
```
I encourage to use the flb_sds_cat_safe() version which takes care of memory reallocation, otherwise by using flb_sds_cat() you will have to check every single return

adiforluls replied:
Fixed

edsiper commented on plugins/out_oci_logan/oci_logan.c:
```diff
@@ -0,0 +1,1327 @@
+
+    flb_http_add_header(c, header, headersize, val, val_size);
+
+    signing_str = flb_sds_cat(signing_str, "\n", 1);
```
use flb_sds_cat_safe()

adiforluls replied:
Fixed in all the places

edsiper commented on plugins/out_oci_logan/oci_logan.c:
```diff
@@ -0,0 +1,1327 @@
+        goto error_label;
+    }
+
+    // Add (requeset-target) to signing string
```
https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#single-line-comment

adiforluls replied:
Fixed in all the places

edsiper commented on plugins/out_oci_logan/oci_logan.c:
```diff
@@ -0,0 +1,1327 @@
+        flb_errno();
+        goto error_label;
+    }
+    signing_str = flb_sds_cat(signing_str, FLB_OCI_HEADER_REQUEST_TARGET,
```
flb_sds_cat_safe()

edsiper commented on plugins/out_oci_logan/oci_logan.c:
```diff
@@ -0,0 +1,1327 @@
+    signing_str = flb_sds_cat(signing_str, encoded_uri,
+                               flb_sds_len(encoded_uri));
+
+    // Add Host to Header
```
https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#single-line-comment

edsiper commented on plugins/out_oci_logan/oci_logan.c:
```diff
@@ -0,0 +1,1327 @@
+    // Add Host to Header
+    if (((c->flags & FLB_IO_TLS) && c->port == 443)
+        || (!(c->flags & FLB_IO_TLS) && c->port == 80)) {
+        // default port
```
https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#single-line-comment

edsiper commented on plugins/out_oci_logan/oci_logan.c:
```diff
@@ -0,0 +1,1327 @@
+        goto error_label;
+    }
+
+    // Add Date header
```
https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#single-line-comment

edsiper commented on plugins/out_oci_logan/oci_logan.c:
```diff
@@ -0,0 +1,1327 @@
+            && (tmp_len == sizeof(FLB_OCI_ERROR_CODE_RELATED_RESOURCE_NOT_FOUND) - 1)
+            && strncasecmp(error_response->code, FLB_OCI_ERROR_CODE_RELATED_RESOURCE_NOT_FOUND, tmp_len) == 0) {
+            ret = FLB_TRUE;
+        } else if( c->resp.status == 401
```
https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#braces-usage-on-conditionals-loops-and-functions

adiforluls replied:
Fixed

edsiper commented on plugins/out_oci_logan/oci_logan.c:
```diff
@@ -0,0 +1,1327 @@
+            &&( tmp_len == sizeof(FLB_OCI_ERROR_CODE_NOT_AUTHENTICATED)-1 )
+            && strncasecmp(error_response->code, FLB_OCI_ERROR_CODE_NOT_AUTHENTICATED, tmp_len) == 0) {
+            ret = FLB_TRUE;
+        } else if (c->resp.status == 404
```
https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#braces-usage-on-conditionals-loops-and-functions

edsiper commented on plugins/out_oci_logan/oci_logan.c:
```diff
@@ -0,0 +1,1327 @@
+    }
+    flb_http_allow_duplicated_headers(c, FLB_FALSE);
+
+    flb_plg_info(ctx->ins, "built client");
```
it seems an info message its a bit verbose for the user, we usually get requirements around reducing the unnecessary logging, maybe you can do some flb_plg_debug()

adiforluls replied:
Yea, these were for debugging so turned them to debug

edsiper commented on plugins/out_oci_logan/oci_logan.c:
```diff
@@ -0,0 +1,1327 @@
+        flb_plg_error(ctx->ins, "failed to build headers");
+        goto error_label;
+    }
+    flb_plg_info(ctx->ins, "built request");
```
it seems an info message its a bit verbose for the user, we usually get requirements around reducing the unnecessary logging, maybe you can do some flb_plg_debug()

adiforluls replied:
fixed

edsiper commented on plugins/out_oci_logan/oci_logan.c:
```diff
@@ -0,0 +1,1327 @@
+            count++;
+        }
+
+        // msgpack_pack_str(&mp_pck, map.via.map.ptr->val.via.str.size);
```
cleanup

adiforluls replied:
fixed

edsiper commented on plugins/out_oci_logan/oci_logan.c:
```diff
@@ -0,0 +1,1327 @@
+    .config_map     = config_map,
+
+    /* Events supported */
+    .event_type   = FLB_OUTPUT_LOGS | FLB_OUTPUT_TRACES,
```
have you tested the plugin and endpoint with traces ?

adiforluls replied:
My bad, this is not supposed to support traces. Fixed.

adiforluls commented:
@edsiper thanks for the review, I've addressed the comments.

edsiper commented:
Please add these 2 missing things to get this sorted out:

- Valgrind output: valgrind --leak-check=full fluent-bit ...
- Link to PR for docs (fluent/fluent-bit-docs)

note: we are cutting a new release on Aug 30, would be great to have this done by then

adiforluls commented:
@edsiper thanks, will share the doc PR soon. 
Also on Valgrind, I'm unable to install valgrind on macOS Ventura, is there any macOS alternative/hack to evaluate memory leaks?

adiforluls commented:
@edsiper For detecting leaks I used "leaks" tool on macOS (Valgrind is unsupported on Ventura). Some leaks were reported and I fixed them. Here's the latest output
```
Process:         fluent-bit [91251]
Path:            /Users/USER/*/fluent-bit
Load Address:    0x103067000
Identifier:      fluent-bit
Version:         0
Code Type:       X86-64
Platform:        macOS
Parent Process:  bash [38805]

Date/Time:       2023-08-30 12:27:45.276 +0530
Launch Time:     2023-08-30 12:27:27.911 +0530
OS Version:      macOS 13.3.1 (22E261)
Report Version:  7
Analysis Tool:   /usr/bin/leaks

Physical footprint:         2644K
Physical footprint (peak):  2644K
Idle exit:                  untracked
----

leaks Report Version: 4.0
Process 91251: 8519 nodes malloced for 919 KB
Process 91251: 0 leaks for 0 total leaked bytes.
```
Let me know if this will suffice. Will be sharing the doc PR in a few hours.

adiforluls commented:
Here's the documentation PR https://github.com/fluent/fluent-bit-docs/pull/1186 
@edsiper

edsiper commented:
@adiforluls is it the "oci_logan" name a recognizable term that describes Oracle Cloud Log Analytics platform ?, I just want to make sure people understand what it means, or if we need to rename it before merging it

patrick-stephens commented on CMakeLists.txt:
```diff
@@ -255,6 +255,7 @@ option(FLB_OUT_PROMETHEUS_REMOTE_WRITE "Enable Prometheus remote write plugin"
 option(FLB_OUT_S3                      "Enable AWS S3 output plugin"                  Yes)
 option(FLB_OUT_VIVO_EXPORTER           "Enabel Vivo exporter output plugin"           Yes)
 option(FLB_OUT_WEBSOCKET               "Enable Websocket output plugin"               Yes)
+option(FLB_OUT_OCI_LOGAN               "Enable OCI Logging analytics plugin"          Yes)
```
Per the docs PR, I would expand OCI to be Oracle Cloud Infrastructure as it is a fairly generic acronym and may be confused with open container initiative.

edsiper replied:
what about `oracle_log_analytics` ? azure has `azure_logs ingestion`

patrick-stephens replied:
Yeah sounds good to me, making it clear it is Oracle I think is the key one - we have AWS/Google equivalents too.

adiforluls replied:
Good idea, I'll make these changes

adiforluls replied:
Refactored @edsiper

adiforluls commented:
Hi @edsiper @patrick-stephens , I see this was added to v2.1.9 milestone but wasn't merged to be included in 2.1.9. Let me know what's holding the merge such that this goes out in the next release.

adiforluls commented:
Thanks @patrick-stephens for you response. 
Below is a summary of comments/changes requested that I've addressed

- The code changes that were requested have been made
- Evaluation of leaks was done and leaks that were found were fixed
- Documentation PR for the plugin was raised
- Renamed the plugin to `oracle_log_analytics`

cc: @edsiper 

Note: Once this is merged, I'll be raising another PR to add two other modes of authentication to this plugin, the size of that PR is roughly the same as this one. It'll be good if we can get this in by next week if the PR looks good such that we have enough time to review the follow up PR and get that in before the next release as well.

edsiper commented:
the plugin build option still has the old name: FLB_OUT_OCI_LOGAN   , same as the registration structure, please get it fixed so we can merge it, thank you

adiforluls commented:
Fixed the old name references @edsiper

edsiper commented on plugins/out_oracle_log_analytics/oci_logan.h:
```diff
@@ -0,0 +1,223 @@
+#define FLB_OCI_ERROR_CODE_TOO_MANY_REQUESTS               "TooManyRequests"
+#define FLB_OCI_ERROR_CODE_INTERNAL_SERVER_ERROR           "InternalServerError"
+
+#define METADATA_URL_BASE  "http://169.254.169.254/opc/v2"
```
why these endpoints are hard coded and there is no host/DNS ? are these definitions in use ?

adiforluls replied:
These are not in use currently but will be in use in the follow up PR to add two other modes of authentication to connect to Oracle Cloud Infrastructure endpoints. I can remove it from this PR if you want.

The reason for no host/DNS is probably because these endpoints were set this way. OCI has SDKs in various languages and they include these endpoints hardcoded just like this.

adiforluls replied:
Removed those endpoints from this PR since they are unused currently.
